### PR TITLE
Avoid creation of component when setting a blank configuration to non…

### LINKED ISSUE #20970: org.keycloak.userprofile.UserProfileProvider is bounced on every realm change
Originally reported in [KEYCLOAK-19044](https://issues.redhat.com/browse/KEYCLOAK-19044):

Whenever the realm is changed:

- the org.keycloak.userprofile.UserProfileProvider is added to the export/components if it was previously missing
- the org.keycloak.userprofile.UserProfileProvider is removed from the export/components if it was previously present
So the org.keycloak.userprofile.UserProfileProvider's presence is constantly bounced when the realm is changed.

Steps to reproduce:

- Start the docker image with: docker run -p 8080:8080 -e KEYCLOAK_USER=keycloak -e KEYCLOAK_PASSWORD=keycloak -e JAVA_OPTS="-Dkeycloak.profile.feature.scripts=enabled -Dkeycloak.profile.feature.upload_scripts=enabled" quay.io/keycloak/keycloak:15.0.1
- In the browser go to the localhost:8080
- Log in with keycloak/keycloak
- Create a new realm - ex 'jose'
- export the realm - the org.keycloak.userprofile.UserProfileProvider is not present
- Change the realm (ex set description to 'a')
- export the realm - the org.keycloak.userprofile.UserProfileProvider is present
- Change the ream (ex set description to 'b')
- export the realm - the org.keycloak.userprofile.UserProfileProvider is not present
- Change the realm (ex set the descriptio to 'c')
- export the realm - the org.keycloak.userprofile.UserProfileProvider is present
- ...

I was able to reproduce the issue with current main.

### PR DESCRIPTION
**Pull request by vramik**
…-existent UserProfileProvider upon realm change

Closes #20970

### CODE REVIEW

**Comment by ahus1** [comment]
@vramik - thank you for this PR. Unfortunately I see several failing tests. Please have a look. I didn't have a look why those fail. If you think this is difficult to fix and applying the Hibernate workaround as described in #22876 is simpler to implement, feel free to use the solution suggested in #22876 instead.

**Review by pedroigor — COMMENTED** [review]
Perhaps we should create the component with the config and avoid calling update.

**Comment by pedroigor** [comment]
@vramik I was looking at it more and one of the reasons we are creating the model is to leverage the notes from the component model so that we avoid parsing the UP config all the time.

Without it, we would need to think how we can avoid unnecessary parsing.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/CustomUserProfileTest.java [hunk 1]
```diff
@@ -52,10 +54,12 @@ public void testCustomUserProfileProviderIsActive() {
     }
 
     private static void testCustomUserProfileProviderIsActive(KeycloakSession session) {
-        DeclarativeUserProfileProvider provider = getDynamicUserProfileProvider(session);
+        UserProfileProvider provider = getUserProfileProvider(session);
         assertEquals(CustomUserProfileProvider.class.getName(), provider.getClass().getName());
         assertTrue(provider instanceof  CustomUserProfileProvider);
-        assertEquals("custom-user-profile", provider.getComponentModel().getProviderId());
+        Optional<ComponentModel> component = getComponentModel(session);
```
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/CustomUserProfileTest.java:60]: @vramik We need to make sure the configuration is set before checking the component. We can add this:

```
provider.setConfiguration(UPConfigUtils.readDefaultConfig());
```

File: services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java [hunk 2]
```diff
@@ -151,16 +154,21 @@ protected UserProfileMetadata configureUserProfile(UserProfileMetadata metadata,
             return decoratedMetadata;
         }
 
-        ComponentModel model = getComponentModelOrCreate(session);
-        Map<UserProfileContext, UserProfileMetadata> metadataMap = model.getNote(PARSED_CONFIG_COMPONENT_KEY);
+        ComponentModel component = getComponentModel().orElse(null);
+
+        if (component == null) {
+            return DEFAULT_METADATA.computeIfAbsent(context, (c) -> decorateUserProfileForCache(decoratedMetadata, defaultRawConfig));
```
Comment (ahus1) [inline services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java:160]: This `computeIfAbsent` feeds data into a static final map `DEFAULT_METADATA`. The dependent data is from what I see dependent on the realm and user, and I don't see how much `DEFAULT_METADATA` can grow here, and how old entries are eventually cleared.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Hi Vlasta - thank you for this change. I had a look, and I have some concerns about `DEFAULT_METADATA` being `static final` here. 

Please review the code. I'd be happy to join a 1:1 meeting to discuss this - please schedule as it fits your calendar.

File: services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java (hunk 2 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java:160]: @ahus1 I made most of these changes on top of @vramik work. You're right and I'll review this part. Not sure if the usages of ream there still make sense.

**Review by pedroigor — APPROVED** [review]
@ahus1 Do you mind reviewing it again?

**Review by ahus1 — APPROVED** [review]
When reviewing this change with @pedroigor, the part `private static final DEFAULT_METADATA` raised my concerns. I now understand that this is a finite list with values mostly hardcoded, and some being based on the Keycloak runtime configuration. The size is therefore finite. 

As that cache would live longer than the Keycloak Session Factory, this will could IMHO cause problems in the model tests where one JVM initializes multiple session factories. Taking to @pedroigor the parts I see problematic here will be changed in the upcoming PR to be either removed or not be static.

**Pull request merged**